### PR TITLE
[Feature] Route optimization for Delivery Trip stops

### PR DESCRIPTION
**Pull request by Alchez**
Re-opening #15390.

<hr>

This PR introduces provisions for Delivery Trip route optimization as described in general [vehicle-routing](https://en.wikipedia.org/wiki/Vehicle_routing_problem) and [travelling salesman](https://en.wikipedia.org/wiki/Travelling_salesman_problem) problems, and the implementation is based on Google's [Directions API](https://developers.google.com/maps/documentation/directions/intro).

**Features:**
- Process and optimize the entire Delivery Trip route, with or without locks
- Form sets of routes for optimization based on applied lock positions
- Re-arrange stops based on the optimized routes and lock positions
- Set delay times between stops - offset estimation times by delays cumulatively
- Add a default distance unit in Global Defaults for system-wide use
- Store distances returned by Google Maps in the Delivery Trip and Stop
- The destination's latitude and longitude are also stored in hidden fields inside Delivery Stop.

<hr>

<details>
<summary> Screenshots / GIFs </summary>

**Without position locks:**

without_locks

**With position locks:**

with_locks

**Global Defaults**

image

</details>

### CODE REVIEW

**Comment by sagarvora** [comment]
Hi @Alchez, your travis build has errored. Also, kindly review #15565 for changes in your implementation made by me.

**Comment by Alchez** [comment]
@sagarvora, oh hey, just saw this. I'll check out the PR - thanks for the update.

**Comment by rmehta** [comment]
Looks good, but needs tests!

https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist#test-cases

File: erpnext/stock/doctype/delivery_trip/delivery_trip.py [hunk 1]
```diff
@@ -45,137 +54,248 @@ def update_delivery_notes(self, delete=False):
...
+	def process_route(self, optimize):
+		"""
+		Estimate the arrival times for each stop in the Delivery Trip.
+		If `optimize` is True, the stops will be re-arranged, based
+		on the optimized order, before estimating the arrival times.
+
+		Args:
+			optimize (bool): True if route needs to be optimized, else False
+		"""
+
```
Comment (rmehta) [inline erpnext/stock/doctype/delivery_trip/delivery_trip.py]: This should work only in the case if Google Maps is enabled

Comment (Alchez) [reply erpnext/stock/doctype/delivery_trip/delivery_trip.py]: Added a check.

**Comment by nabinhait** [comment]
Travis is failing.

**Comment by Alchez** [comment]
@nabinhait, I've fixed the tests. Although Travis is now failing at trying to migrate `User Permission`.

**Pull request merged**